### PR TITLE
Add support for pattern validation annotation

### PR DESCRIPTION
**Pull request by bezmax**
Adds support for `@Pattern` validation annotation. 

Also refactored the structure of validation tests by separating tests for different types of validator implementations. In the future it would be best to also separate them into separate files - one test file per validator implementation.

Fixes canoo/dolphin-platform#659

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/670)

### CODE REVIEW

**Comment by hendrikebbers** [comment]
Thx :)
Added some first comments. Will have a look at the test class tomorrow

**Comment by hendrikebbers** [comment]
@bezmax only some small changes needed here :) very good work. Can you fix it based on my comments? Will merge it once this is done

**Comment by bezmax** [comment]
@hendrikebbers 
> Can you fix it based on my comments?

Sorry, I'm not sure I know where to find those comments :D

**Comment by hendrikebbers** [comment]
Just scroll up ^^^I added some comments to code snippets of your PR. Or go to the "Files changed" tap on the top of the PR :)

**Comment by bezmax** [comment]
@hendrikebbers Ok, I'm either not seeing it, or there is maybe some permission problem not allowing me to see your comments. Here's what I see if I go to "Files changed" on top of PR: https://imgur.com/a/jXoRM

Also, I did not receive any emails notifying you about your comments on my PR, which I did from other repo I created a PR to.

**Comment by hendrikebbers** [comment]
@bezmax now?

**Comment by bezmax** [comment]
@hendrikebbers Much better now, thanks.

The 'minValue' one is kind of embarrassing :D Going to fix those in a sec.

File: platform-extras/dolphin-platform-bean-validation/src/main/java/com/canoo/dp/impl/validation/PatternPropertyValidator.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        int combined = 0;
+        for (Pattern.Flag f : flags) {
+            combined |= f.getValue();
+        }
+        return combined;
+    }
+
+    @Override
+    protected boolean checkValid(
+            @NotNull CharSequence value,
```
[the remark this answers is not in the document]
Comment (bezmax) [reply platform-extras/dolphin-platform-bean-validation/src/main/java/com/canoo/dp/impl/validation/PatternPropertyValidator.java]: `@NotNull` is used for static code analysis. I have noticed from `AbstractPropertyValidator` implementation that the value is never null in that method (`onNullValue` is used instead), so I added it onto the parameter. Now, for example, if you try to do defensive programming and add another null check in the method - you will get a warning. Also, if `AbstractPropertyValidator` is ever changed to pass null values to that method - FindBugs will warn about trying to pass potential null into a `@NotNull` parameter.

File: platform-extras/dolphin-platform-bean-validation/src/main/java/com/canoo/dp/impl/validation/PatternPropertyValidator.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    /**
+     * constructor
+     */
+    public PatternPropertyValidator() {
+        super(CharSequence.class);
+    }
+
+    @Override
+    public void initialize(Pattern minValue) {
+        int flags = combineFlags(minValue.flags());
```
[the remark this answers is not in the document]
Comment (bezmax) [reply platform-extras/dolphin-platform-bean-validation/src/main/java/com/canoo/dp/impl/validation/PatternPropertyValidator.java]: I'm curious, given that Guava is used in the project, why not use guava's `Preconditions` class?

**Comment by hendrikebbers** [comment]
I hope that we can remove guava in future ;). Having less dependencies is always better for a framework.

Von meinem iPhone gesendet

**Comment by bezmax** [comment]
@hendrikebbers All requested changes are done.

**Comment by hendrikebbers** [comment]
Cool! THX :)

**Pull request merged**